src/framework-ondata.js

```diff
@@ -1,18 +1,164 @@
- * with the real version of the code once the pipeline-component-factory is finalized.
- *
+ * The standard RDF pipeline framework ondata method. This is called whenever  
+ * there is new data on any port.  This method will update the input state,
+ * and determine if the component updater should be executed now or not. 
+ * If the updater should be executed, the fRunUpdater wrapper API will be
+ * called to to handle that.  The updater output (or error) will be stored 
+ * in output or error state, and those results will be sent on to any 
+ * down stream components for their processing.
+ * 
  * @this port context
- *
- * @param payload data payload on the input port
- * @param socketIndex (optional) this will be set only for ports configured for
- *        multiple inputs (multi/noflo addressable).
+ * @param payload data coming into the port
+ * @param socketIndex socket index if the port is a multiple input (addressable) 
  */
 module.exports = function( payload, socketIndex ) {
 
-     return;
+     var portName = this.name;
+
+     var vnid = payload.vnid || '';
+     var vni = this.nodeInstance.vni(vnid);
+
+     // Update the input state
+     var inputState = (payload.vnid) ? payload : stateFactory( vnid, payload );
+     vni.inputStates( portName, socketIndex, inputState );
+
+     // check if should run updater & run if necessary
+     if ( shouldRunUpdater( vni ) ) { 
+          
+        // Save vars we will need in the promise where the context is different
+        var outputPorts = this.nodeInstance.outPorts;
+
+        if ( _.isUndefined( this.nodeInstance.wrapper.fRunUpdater ) || 
+             ! _.isFunction( this.nodeInstance.wrapper.fRunUpdater ) ) { 
+
+             // Don't have a wrapper fRunUpdater 
+             throw Error( 'No wrapper fRunUpdater function found!  Cannot run updater.' );
+
+        } else { 
+
+            // save current error & output state LMs to compare after updater runs
+            var lastOutputLm = vni.outputState().lm;
+            var lastErrorLm = vni.errorState().lm;
+            var wrapper = this.nodeInstance.wrapper;
+
+            new Promise(function( resolve ) { 
+                resolve( wrapper.fRunUpdater.call( vni ) );
+
+            }).then( function() { 
+                // If not new output or error state data, send it downstream or log it to the console
+                handlePortOutput( outputPorts.output, lastOutputLm, vni.outputState() );
+                handlePortOutput( outputPorts.error,  lastErrorLm, vni.errorState() );
+
+            }, function( rejected ) { 
+    
+                console.error( rejected );
+
+                // fRunUpdater failed - is this a new error? 
+                var errorState = vni.errorState();
+
+                if ( lastErrorLm !== errorState.lm  ) { 
+
+                    // fRunUpdater or updater recorded a new lm (i.e., new error) - send it on
+                    handlePortOutput( outputPorts.error, lastErrorLm, errorState );
```

Comment:
Good point.  And it raises the issue of what assumptions we should make about the content of error.data , and how we can distinguish between warnings and errors.  We need to distinguish because when an upstream node is in an error state then the downstream node's updater must not fire.  We should discuss how to do this.  Maybe consider a node to be in an error state if error.data is a string and contains the word "error", or if error.data is an object of type Error?